core/schema/git.go

```diff
@@ -47,125 +44,86 @@ func (s *gitSchema) Dependencies() []ExecutableSchema {
 	return nil
 }
 
-type gitRepository struct {
-	URL             string         `json:"url"`
-	KeepGitDir      bool           `json:"keepGitDir"`
-	AuthTokenSecret *core.SecretID `json:"authTokenSecret,omitempty"`
-	Pipeline        pipeline.Path  `json:"pipeline"`
-	ServiceHost     *core.Service  `json:"serviceHost,omitempty"`
-}
-
-type gitRef struct {
-	Repository gitRepository
-	Name       string
-}
-
 type gitArgs struct {
 	URL                     string          `json:"url"`
 	KeepGitDir              bool            `json:"keepGitDir"`
 	ExperimentalServiceHost *core.ServiceID `json:"experimentalServiceHost"`
+
+	SSHKnownHosts string    `json:"sshKnownHosts"`
+	SSHAuthSocket socket.ID `json:"sshAuthSocket"`
```

Comment:
Sorry for only seeing this now, it's good that these have moved out of `tree` but wouldn't it be better to put it in a `withXXX` method? Back in July I tried adding token auth but couldn't get it to work even by doing the same thing buildkit does underneath.

So these would be different authentication methods:

```graphql
  withAuthToken(secret: SecretID!): GitRepository!
  withAuthHeader(secret: SecretID!): GitRepository!
  withSSHAuth(socket: SocketID!, knownHosts: String): GitRepository!
```

Reply:
Agh, sorry, I should've added you as a reviewer on this.

I kind of agree, the `with` methods probably make more sense for this. Unfortunately, since this change shipped in https://github.com/dagger/dagger/releases/tag/v0.9.1, we'd need to deprecate the top-level constructor methods. That seems fine to me - but does feel odd - maybe doing this ASAP makes most sense, so no one has a chance to actually use the new fields?

Tiny bikeshed on the auth methods: for `withSSHAuth`, `socket` and `knownHosts` are not related to each other, it's possible to update `knownHosts` without updating the `socket`, so we might want to split those to two separate methods (or at least, `SocketID` also needs to not be required).

Reply:
> [...] `socket` and `knownHosts` are not related to each other, it's possible to update `knownHosts` without updating the `socket` [...]

Oh, you're right.